### PR TITLE
Add EFA support for AWS

### PR DESCRIPTION
**Pull request by cmd-ntrf**

### CODE REVIEW

**Comment by ocaisa** [comment]
The interesting thing is, it works if I add the "efa" tag to `mgmt1` and `login1`.

**Comment by cmd-ntrf** [comment]
I am making the hypothesis that we might need a dedicated subnet containing only EFA enabled instances.

I have added a new subnet in the VPC and EFA network interface only in it. Let me know if it works.

**Comment by ocaisa** [comment]
Unfortunately, that doesn't work, on the EFA nodes I get:
```
Jul 13 19:46:48 node2 puppet-agent[1700]: (/Stage[main]/Profile::Freeipa::Client/File[/sbin/mc-ipa-client-install]) Could not evaluate: Could not retrieve file metadata for puppet:///modules/profile/freeipa/mc-ipa-client-install: Request to https://puppet:8140/puppet/v3/file_metadata/modules/profile/freeipa/mc-ipa-client-install?links=manage&checksum_type=md5&source_permissions=ignore&environment=production failed after 0.001 seconds: Failed to open TCP connection to puppet:8140 (No route to host - connect(2) for "puppet" port 8140)
```

**Comment by cmd-ntrf** [comment]
Ok, so it means there is no default route between the subnet. I'll look how we can define one.

In the meantime, is it possible to determine if EFA works without configuring the entire node with puppet?

**Comment by ocaisa** [comment]
Puppet still seems to be struggling on, I think it may still register in the end

**Comment by ocaisa** [comment]
Ah no, NFS is also falling over:
```
Jul 13 20:14:13 node1 mount[6568]: mount to NFS server '10.0.0.52' failed: No route to host, retrying
```
I'll have to come back to this tomorrow.

**Comment by ocaisa** [comment]
The issue seems to be the firewall rules within MC. After checking `iptables`, I saw these are currently set to `10.0.0.0/24` which will block traffic between subnets. After I changed the rule on the management node, a lot of the problems disappeared...although I still see
```
Jul 14 08:45:31 node2 puppet-agent[1690]: (/Stage[main]/Profile::Nfs::Client/Nfs::Client::Mount[/scratch]/Mount[shared /scratch by 10.0.0.93 on /scratch]) Could not evaluate: Execution of '/usr/bin/mount /scratch' returned 32: mount.nfs4: access denied by server while mounting 10.0.0.93:/scratch
```

**Comment by ocaisa** [comment]
I hacked the `getcidr()` function and things succeeded, but unfortunately we end with the same situation as before, the EFA fabric does not seem to be working. I'll see if I can reach out to someone in AWS.

**Comment by ocaisa** [comment]
I tried adding EFA as a second interface on the compute nodes (with their own security group) but in that case the nodes failed to connect to  a yum repo.

**Comment by cmd-ntrf** [comment]
Ok so the subnet does not fix anything. I have reverted the commit.

Next time you test this with login and mgmt1 not having the efa tag, can you verify the EFA adapter is attached on a compute node with :
```
lspci | grep efa0
```

**Comment by ocaisa** [comment]
`efa0` is not there but the EFA adaptor is connected:
```
[EESSI pilot 2021.03] $ lspci
00:00.0 Host bridge: Intel Corporation 440FX - 82441FX PMC [Natoma]
00:01.0 ISA bridge: Intel Corporation 82371SB PIIX3 ISA [Natoma/Triton II]
00:01.3 Non-VGA unclassified device: Intel Corporation 82371AB/EB/MB PIIX4 ACPI (rev 08)
00:03.0 VGA compatible controller: Amazon.com, Inc. Device 1111
00:04.0 Non-Volatile memory controller: Amazon.com, Inc. Device 8061
00:05.0 Ethernet controller: Amazon.com, Inc. Elastic Network Adapter (ENA)
00:06.0 Ethernet controller: Amazon.com, Inc. Elastic Fabric Adapter (EFA)
```

**Comment by ocaisa** [comment]
Final test for today was to create a working EFA cluster, and then simply remove the `efa` tag of login1 and apply. This alone is enough to trigger the performance drop. Whatever is causing this is very subtle. My suspicion is that we are tainting the security group by including a non-efa nic in there, but that is just a guess.

**Comment by ocaisa** [comment]
Ok, I've found the root of the problem and it is not in the infrastructure setup. I tested including the login node in the EFA setup and excluding the management node, this gave me performing EFA fabric. Removing the login node from the EFA lowers the performance. After that I suspected that it must be something to do with either SLURM or OpenMPI and not with AWS infrastructure. I ran a job which included some debug information:
<details>
<summary>MPI debug jobscript</summary>

```
#!/bin/bash -x
#SBATCH --time=00:20:00
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=1
#SBATCH --nodes=2

module load OSU-Micro-Benchmarks

ldd $(which osu_latency)

# Check SLURM envvars
env | grep SL
# Check SBATCH envvars
env | grep SB

mpirun -mca mpi_show_mca_params all -n 2 osu_latency
mpirun -n 2 osu_bw
```
</details>

This lead me to discover that OpenMPI is behaving differently in the different scenarios:

<details>
<summary>Filtered diffs of output from different scenarios</summary>

```
[ocaisa@login1 ~]$ diff working_efa.out non-working_efa.out 
159c159
< [node1.int.aws.learnhpc.eu:] orte_jobfam_session_dir=/tmp/ompi.node1.60005/pid.7189 (environment)
---
> [node1.int.aws.learnhpc.eu:] orte_jobfam_session_dir=/tmp/ompi.node1.60005/pid.21827 (environment)
228c228
< [node1.int.aws.learnhpc.eu:] orte_ess_jobid=1467809793 (environment)
---
> [node1.int.aws.learnhpc.eu:] orte_ess_jobid=506134529 (environment)
318a319,330
> [node1.int.aws.learnhpc.eu:] pml_ob1_verbose=0 (default)
> [node1.int.aws.learnhpc.eu:] pml_ob1_free_list_num=4 (default)
> [node1.int.aws.learnhpc.eu:] pml_ob1_free_list_max=-1 (default)
> [node1.int.aws.learnhpc.eu:] pml_ob1_free_list_inc=64 (default)
> [node1.int.aws.learnhpc.eu:] pml_ob1_priority=20 (default)
> [node1.int.aws.learnhpc.eu:] pml_ob1_send_pipeline_depth=3 (default)
> [node1.int.aws.learnhpc.eu:] pml_ob1_recv_pipeline_depth=4 (default)
> [node1.int.aws.learnhpc.eu:] pml_ob1_max_rdma_per_request=4 (default)
> [node1.int.aws.learnhpc.eu:] pml_ob1_max_send_per_range=4 (default)
> [node1.int.aws.learnhpc.eu:] pml_ob1_unexpected_limit=128 (default)
> [node1.int.aws.learnhpc.eu:] pml_ob1_use_all_rdma=false (default)
> [node1.int.aws.learnhpc.eu:] pml_ob1_allocator=bucket (default)
322,334d333
< [node1.int.aws.learnhpc.eu:] pml_cm_free_list_num=4 (default)
< [node1.int.aws.learnhpc.eu:] pml_cm_free_list_max=-1 (default)
< [node1.int.aws.learnhpc.eu:] pml_cm_free_list_inc=64 (default)
< [node1.int.aws.learnhpc.eu:] mtl= (default)
< [node1.int.aws.learnhpc.eu:] mtl_base_verbose=error (default)
< [node1.int.aws.learnhpc.eu:] mtl_ofi_priority=25 (default)
< [node1.int.aws.learnhpc.eu:] mtl_ofi_provider_include= (default)
< [node1.int.aws.learnhpc.eu:] mtl_ofi_provider_exclude=shm,sockets,tcp,udp,rstream (default)
< [node1.int.aws.learnhpc.eu:] mtl_ofi_progress_event_cnt=100 (default)
< [node1.int.aws.learnhpc.eu:] mtl_ofi_tag_mode=auto (default)
< [node1.int.aws.learnhpc.eu:] mtl_ofi_control_progress=unspec (default)
< [node1.int.aws.learnhpc.eu:] mtl_ofi_data_progress=unspec (default)
< [node1.int.aws.learnhpc.eu:] mtl_ofi_av=map (default)
446,469c445,468
```
</details>

So basically for some reason if EFA is not on the login node OpenMPI does do try to use libfabric (which is where the EFA support lives).

**Comment by ocaisa** [comment]
I'm no OpenMPI expert and I can't seem to force it to use libfabric in the poor performance scenario. Maybe someone else knows what this might happen?

**Comment by cmd-ntrf** [comment]
Could this be related to environment variables set on the login node during openmpi module load and propagated to the compute node by Slurm?

Can you try setting the Slurm flag `#SBATCH --export=NONE` and adding the module load of openmpi to your job?

**Comment by ocaisa** [comment]
> Could this be related to environment variables set on the login node during openmpi module load and propagated to the compute node by Slurm?
> 
> Can you try setting the Slurm flag `#SBATCH --export=NONE` and adding the module load of openmpi to your job?

I can try it tomorrow...but I already have the load in the jobscripts and typically didn't have it loaded at job submission. Perhaps exporting no envvars has an impact somehow though.

**Comment by ocaisa** [comment]
I tried `#SBATCH --export=NONE` and it had no effect. I even tried job submission from JupyterHub, which means I submitted from EFA-enabled node, this also did not use libfabric. I really don't know why this is happening. Is `login1` special somehow when it comes to SLURM perhaps?

**Comment by ocaisa** [comment]
Ok, we tracked this down a little. It would seem that if the node you ssh into has EFA, then the job starts with EFA supported. If it does not then the job fails to use the EFA device due to:
```
[node2.int.aws2.learnhpc.eu:08544] mtl_ofi_component.c:710: fi_enable failed: Cannot allocate memory
```
The only things in the environment related to the login node when MPI is invoked are `SSH_CONNECTION` and `SLURM_SUBMIT_HOST`. Using `--export=None` removes the ssh envvar and has no impact. Altering or removing `SLURM_SUBMIT_HOST` has no impact.

One suspicion right now is that this we have `pmi2` in SLURM rather than `pmix`.

File: common/instance_config/puppet.yaml [hunk 1]
```diff
@@ -66,6 +66,12 @@ runcmd:
   - grub2-mkconfig -o /boot/grub2/grub.cfg
 # Remove all ifcfg configuration files that have no corresponding network interface in ip link show.
   - for i in /etc/sysconfig/network-scripts/ifcfg-*; do if ! ip link show | grep -q "$${i##*-}:"; then rm -f $i; fi; done
+# AWS EFA driver installation
+%{ if contains(tags, "efa") }
+  - curl -O https://efa-installer.amazonaws.com/aws-efa-installer-latest.tar.gz
+  - "(tar xf aws-efa-installer-latest.tar.gz && cd aws-efa-installer && ./efa_installer.sh --yes --minimal)"
+  - rm -f aws-efa-installer-latest aws-efa-installer-latest.tar.gz
```
Comment (ocaisa) [inline common/instance_config/puppet.yaml:73]:
```suggestion
  - rm -f aws-efa-installer aws-efa-installer-latest.tar.gz
```

**Comment by ocaisa** [comment]
Well, that's annoying, installing the EFA drivers on the login node (which does not have EFA) is enough to give the performance.

File: common/instance_config/puppet.yaml [hunk 2]
```diff
@@ -66,6 +66,12 @@ runcmd:
   - grub2-mkconfig -o /boot/grub2/grub.cfg
 # Remove all ifcfg configuration files that have no corresponding network interface in ip link show.
   - for i in /etc/sysconfig/network-scripts/ifcfg-*; do if ! ip link show | grep -q "$${i##*-}:"; then rm -f $i; fi; done
+# AWS EFA driver installation
+%{ if contains(tags, "efa") }
```
Comment (ocaisa) [inline common/instance_config/puppet.yaml:70]: I did another test with a mix of EFA and non-EFA nodes. If the EFA drivers are not installed, a simple bandwidth test gives
```
# OSU MPI Bandwidth Test v5.6.3
# Size      Bandwidth (MB/s)
[node1.int.aws2.learnhpc.eu:08156] mtl_ofi_endpoint.h:47: *** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[21688,1],0]
  Exit code:    1
--------------------------------------------------------------------------
```
but if we just install the EFA drivers on the non-EFA nodes we get a much clearer error:
```
[node1.int.aws2.learnhpc.eu:08559] pml_ucx.c:273  Error: Failed to create UCP worker
# OSU MPI Latency Test v5.6.3
# Size          Latency (us)
[node1.int.aws2.learnhpc.eu:08559] mtl_ofi_endpoint.h:47: *** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
[tnode1.int.aws2.learnhpc.eu:02146] [[27148,1],1] selected pml ucx, but peer [[27148,1],0] on node1 selected pml cm
[tnode1.int.aws2.learnhpc.eu:02146] pml_ucx.c:452  Error: Failed to resolve UCX endpoint for rank 0
```
I think it makes sense then to install the EFA drivers if _anyone_ is using EFA.

**Comment by cmd-ntrf** [comment]
After some digging in the RPMs installed by aws-efa-installer.sh, I found what was happening in our setup.

efa-config*.rpm installs /etc/security/limits.d/efa.conf, which configures the memlock limits to unlimited.
The default value for memlock in CentOS 7 and 8 is 64 KB.

Because we did not install the efa RPMs on the head node, its memlock limit was left to 64 KB.
By default, Slurm propagates all security limits from the user's process environment on the submit node.
This corresponds to the slurm.conf parameter "PropagateResourceLimits".

During our tests, we were therefore propagating the 64KB memory limits of the head node to the compute nodes, hence the memory allocation error and the subpar bandwidth.

We will deactivate resource limit propagation from now on by setting PropagateResourceLimits to NONE. I am working on a PR for puppet-magic_castle.

The problem is also explained in Slurm FAQ, we were just not searching there for answers:
https://slurm.schedmd.com/faq.html#memlock

File: common/instance_config/puppet.yaml (hunk 2 above)
Comment (cmd-ntrf) [reply common/instance_config/puppet.yaml:72]: Do you think it still makes sense to install it everywhere now that we found the problem was the memlock security limits?

Comment (ocaisa) [reply common/instance_config/puppet.yaml:72]: No, I think either updates the limits on login nodes or adjust the SLURM configuration is all that's required. Will test both tomorrow.
